### ISSUE TITLE
Question about loss function in finetune.py - missing l_3component in total_loss?

### ISSUE THREAD

**Issue by zjh529** [SeekingDream/DLCompilerAttack#2]
Hello Authors,

Thank you for sharing this excellent paper and the accompanying code. The DCL-BD attack method is very insightful.

I was reviewing the Stage 2 implementation in `src/attack/finetune.py` and had a question about the loss function being optimized, as it seems to differ slightly from the paper.

**In the paper (Section 5.4, Eq. 7):**
The total loss `l` is defined as the sum of all four components:
$l = l_1 + l_2 + l_3 + l_4$

**In the code (`src/attack/finetune.py`):**

1.  The `compute_loss` function correctly identifies the components, for example:

      * `loss_cl_D_cl` (as $l_1$)
      * `loss_bd_D_cl` (as $l_2$)
      * `loss_cl_C_cl` (as $l_3$, though it's set to `torch.tensor(0)` in the buggy version)
      * `loss_bd_C_bd` (as $l_4$)

2.  However, the `total_loss` that is actually used for backpropagation in `train_bd_epoch` is taken from `loss[-1]`:

    ```python
    # in train_bd_epoch:
    loss, acc = self.compute_loss(compiled_model, batch)
    total_loss = loss[-1] # <-- This is used for backprop
    optimizer.zero_grad()
    total_loss.backward()
    optimizer.step()
    ```

3.  This `loss[-1]` value is defined in the `return` statement of `compute_loss` as:

    ```python
    # in compute_loss:
    return (
        [..., loss_cl_D_cl + loss_bd_D_cl + loss_bd_C_bd], # <-- loss[-1]
        [...]
    )
    ```

**My Question:**
This `total_loss` appears to correspond to $l_1 + l_2 + l_4$, and it omits the $l_3$ component (`loss_cl_C_cl`).

Even after correcting the known bug in `compute_loss` (i.e., properly calculating `C_cl_embeds`, `C_cl_logit`, and the real `loss_cl_C_cl`), this `total_loss` definition *still* omits the $l_3$ component from the backpropagation step.

Was this omission of $l_3$ from the `total_loss` intentional? Perhaps for stability, or because $l_1$ was found to be a sufficient proxy? Or is this a potential typo in the `return` statement?

Thank you for any clarification you can provide\!

**Comment by SeekingDream** [comment]
Yes, you’re right — this is a simplified version of our paper. We observed that ( l_3 ) and ( l_1 ) behave almost identically during optimization (for clean inputs ( x ), ( M_1 ) and ( C_1 ) produce nearly the same outputs). Therefore, we omitted ( l_3 ) to accelerate training and reduce GPU memory usage. You can include ( l_3 ) if desired; it will not affect the results.

**Comment by SeekingDream** [comment]
We will consider adding it as a user-defined option in a future update. Thank you again for your comment.
